{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BinderHub Snakemake 使用 mock-up\n",
    "\n",
    "## I. Overview\n",
    "\n",
    "### 前言\n",
    "[Depositar](https://data.depositar.io/)為得讓所儲存的資料集達成FAIR data(Findable, Accessible, Interoperable, Reusable)，引進了[BinderHub](https://binderhub.readthedocs.io/en/latest/)來讓資料得以在內部進行運作驗證(interoperable)，而Snakemake的加入則是為了省去複雜的資料驗證程序，並讓資料更新可以只重新執行需要的部分，降低資源使用，而\"省去複雜的資料驗證程序\"的部分則可以彈性使用，讓使用者也可以以自己的資料去驗證實驗結果(Reusable)。\n",
    "\n",
    "這份文件簡單的介紹了Snakemake的使用方式並模擬加入BinderHub\n",
    "\n",
    "### 簡介\n",
    "snakemake 是一個處理工作流程的工具，透過與其他軟件搭配(ex: samtools, bwa …)或單純使用shell, python指令建立各種規則(rule)對資料進行更分析、產生結果報告，類似於Makefile，但有功能使用更使用者友善一些。\n",
    "\n",
    "### 內容概述\n",
    "利用Snakemake來讓作者將其研究方法(分析順序)的Snakefile、script.sh、資料上傳至depositar，當其他使用者帶著自己的資料集來驗證時，上傳至binder中，在終端機中輸入\"bash script.sh\"，系統便會自動下載snakemake與snakedeploy或其他所需要的環境，並直接執行Snakefile，而根據不同需求，Snakefile可能會要取用到其他地方(ex: Github)的資源，或就在本地端執行即可，若想將資料的處理分為前置作業與資料分析，可使用多個Snakefile來達到多層次的工作流程。而Snakemake除了產生出工作流程後產生的成果外，也有內建的功能可以靠指令產生工作流程圖(DAG有向無環圖)或report(.zip,html…)。\n",
    "\n",
    "## II. Snakemake 基本教學\n",
    "\n",
    "### mamba Setup \n",
    "(來源:[安裝流程](https://snakemake.readthedocs.io/en/stable/tutorial/setup.html#step-1-installing-mambaforge))\n",
    "\n",
    "+ Linux 環境\n",
    "```\n",
    "$ curl -L https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-Linux-x86_64.sh -o Mambaforge-Linux-x86_64.sh\n",
    "$ bash Mambaforge-Linux-x86_64.sh\n",
    "```\n",
    "> 下載跟安裝Miniconda 3  \n",
    " \n",
    "![mamba-1](pictures/mamba-1.png)  \n",
    "![mamba-2](pictures/mamba-2.png)  \n",
    "![mamba-3](pictures/mamba-3.png)  \n",
    "\n",
    "> 一直按Enter繼續   \n",
    "\n",
    "![mamba-4](pictures/mamba-4.png)  \n",
    "> yes  \n",
    "\n",
    "![mamba-5](pictures/mamba-5.png)  \n",
    "\n",
    "> 這樣就可以使用mamba了  \n",
    "\n",
    "### 環境設置\n",
    "```\n",
    "# environment.yaml\n",
    "\n",
    "name: environment               # 環境名稱\n",
    "\n",
    "channels:\n",
    "  - conda-forge\n",
    "  - bioconda                    # 生物信息學軟體包的 channel\n",
    "\n",
    "dependencies:                   # 函式庫\n",
    "  - snakemake-minimal >=8.4.4   # 可強調版本\n",
    "  - python\n",
    "  - matplotlib\n",
    "  - pandas\n",
    "  - numpy\n",
    "  - plotly\n",
    "  - folium\n",
    "  - pip                         # 這裡是 Conda 自身的 pip 包\n",
    "\n",
    "  # 如果有需要從 PyPI 安裝的包，可以這樣放在 pip 下  (conda沒有的)\n",
    "  # - pip:\n",
    "  #   - python_package_not_in_conda\n",
    "```\n",
    "\n",
    "### 劇本\n",
    "\n",
    "```\n",
    "# script.sh\n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "# Check if Snakemake is installed, if not, install it\n",
    "if ! command -v snakemake &> /dev/null\n",
    "then\n",
    "    echo \"Snakemake not found, installing...\"\n",
    "    \n",
    "    conda activate base\n",
    "    mamba create -c conda-forge -c bioconda --name snakemake snakemake snakedeploy -y\n",
    "\n",
    "fi\n",
    "\n",
    "# Initialize conda in the script\n",
    "source $(conda info --base)/etc/profile.d/conda.sh\n",
    "\n",
    "conda activate snakemake\n",
    "\n",
    "# Run the Snakemake workflow\n",
    "snakemake  --cores all --use-conda\n",
    "```\n",
    "\n",
    " + 先確認目前有沒snakemake \n",
    "\n",
    "    ->  沒有就安裝\n",
    "\n",
    "\t->  有就啟動環境 \n",
    "\n",
    "  + 因為在建立環境時會問y/n，所以加 -y 回答\n",
    "\n",
    "\n",
    "```\n",
    "if ! command -v snakemake &> /dev/null\n",
    "then\n",
    "    echo \"Snakemake not found, installing...\"\n",
    "    # conda install -n base -c conda-forge mamba\n",
    "    conda activate base\n",
    "    mamba env create -n snakemake --file environment.yaml || mamba env update -n snakemake --file environment.yaml\n",
    "    # mamba create -c conda-forge -c bioconda -n snakemake snakemake\n",
    "fi\n",
    "```\n",
    "ex2(視範例)的用法，不適下載snakemake與snakedeploy，而是安裝environment.yaml裡所需的環境，以 ``` || ``` 來判斷若已有環境，則看是否更新\n",
    "\n",
    "### rules \n",
    "\n",
    "規則範例:\n",
    "\n",
    "```\n",
    "from snakemake.utils import min_version\n",
    "min_version(\"6.10.0\")\n",
    "\n",
    "rule all:\n",
    "    input:\n",
    "        directory(\"data\"),\n",
    "        directory(\"report\")  \n",
    "\n",
    "rule unzip_workflow:\n",
    "    input:\n",
    "        \"workflow.zip\"\n",
    "    output:\n",
    "        \"workflow/Snakefile\"\n",
    "    shell:\n",
    "        \"unzip {input} -d ./\"\n",
    "\n",
    "rule unzip_file:\n",
    "    input:\n",
    "        \"data.zip\",\n",
    "        \"workflow/Snakefile\"\n",
    "    output:\n",
    "        directory(\"data\"),\n",
    "        \"workflow/new_snakefile\"\n",
    "    shell:\n",
    "        '''\n",
    "        unzip {input[0]} -d {output[0]}\n",
    "        mv {input[1]} {output[1]}\n",
    "        '''\n",
    "\n",
    "rule run_workflow:\n",
    "    input: \n",
    "        \"workflow/new_snakefile\"\n",
    "    output:\n",
    "        \"fin_workflow\"\n",
    "    shell: \n",
    "        '''\n",
    "        set +e\n",
    "        snakemake -s {input} --cores all --use-conda\n",
    "        set -e\n",
    "        touch fin_workflow\n",
    "        mv workflow/new_snakefile workflow/Snakefile\n",
    "        '''\n",
    "        \n",
    "rule generate_report:\n",
    "    input:\n",
    "        rules.run_workflow.output\n",
    "    output: \n",
    "        \"report.zip\"\n",
    "    shell:\n",
    "        '''\n",
    "        snakemake -s workflow/Snakefile --report {output}\n",
    "        rm -rf {input}\n",
    "        '''\n",
    "\n",
    "rule unzip_report:\n",
    "    input:\n",
    "        rules.generate_report.output\n",
    "    output:\n",
    "        directory(\"report\")\n",
    "    shell:\n",
    "        \"unzip {input} -d {output}\"\n",
    "```\n",
    "\n",
    "**rule all 是 $ snakemake 的時候跑的**\n",
    "\n",
    "形成下圖流程圖 (順序由上而下)\n",
    "\n",
    "![binder2-dag](pictures/binder2-dag.png)\n",
    "\n",
    "### rule attribute\n",
    "\n",
    "+ input:\n",
    "> 放入這個rule會用到的檔案(可以不用用到的都放，但放較好)，為達成workflow，會放入前一個rule的output來強調順序\n",
    ">\n",
    "> 路徑名需加 **\" \"**\n",
    "> \n",
    "> 多個輸出以 **,** 分開 (ex: “report.zip” **,** directory(“data”))\n",
    "> \n",
    ">有放在input的，在其他地方使用時，就可以用{input}取代，反之，若沒有，則得打出完整路徑(ex: workflow/Snakefile )，有多個input時，在其他地方按順在input的順序以{input[0]},{input[1]}...使用\n",
    "> \n",
    ">指定rule的輸出為input: rules.rule_name,output\n",
    ">\n",
    "\n",
    "+ output:\n",
    "> 放入這個rule會輸出的檔案(可以不用都放，但放較好)，為達成workflow，駛下一個rule可以靠的此rule的output建立順序\n",
    ">\n",
    "> 路徑名需加 **\" \"**\n",
    ">\n",
    "> 多個輸出以 **,** 分開 (ex: “report.zip” **,** directory(“data”))\n",
    ">\n",
    "> 有放在output的，在其他地方使用時，就可以用{output}取代，反之，若沒有，則得打出完整路徑，有多個output時，在其他地方按順在output的順序以{output[0]},{output[1]}...使用\n",
    ">\n",
    "\n",
    "```\n",
    "rule generate_report:\n",
    "    input:\n",
    "        rules.run_workflow.output\n",
    "    output:\n",
    "        \"report.zip\"\n",
    "    shell:\n",
    "        '''\n",
    "        snakemake -s workflow/Snakefile --report {output}\n",
    "        rm -rf {input}\n",
    "        '''\n",
    "```\n",
    "\n",
    "+ shell:\n",
    "> 寫入所要執行的shell指令\n",
    ">\n",
    "> 單個指令用 **\" \"** (ex: “snakemake -s workflow/Snakefile --report {output}”)\n",
    ">\n",
    "> 多個指令用 **''' '''**\n",
    "> \n",
    "> 指令與python程式碼都要用 **\"\"\" \"\"\"** 包住全部， **僅python_code要另外用 python -c “..python code.. “ 在包python程式碼** ，shell指令不用\n",
    "\n",
    "```\n",
    "rule data_update:\n",
    "    input:\n",
    "        rules.deploy.output[0],\n",
    "        rules.deploy.output[1]\n",
    "    output:\n",
    "        \"workflow/new_snakefile\"\n",
    "    shell:\"\"\"\n",
    "        python -c \"\n",
    "import re\n",
    "# Read the file\n",
    " /* more */\n",
    "        \"\n",
    "        mv {input[0]} {output}\t# shell 指令\n",
    "    \"\"\"\n",
    "\n",
    "```\n",
    "+ run:\n",
    ">\n",
    "> 直接放入python程式碼，不用 \" \"\n",
    "> \n",
    "> 加入shell指令用 shell(\"/...shell command.../\")\n",
    ">\n",
    "\n",
    "```\n",
    "run:\n",
    "        with open(input.report) as f:\n",
    "            report_content = f.read()\n",
    "\n",
    "\n",
    "        # 讀取 CSV 文件內容\n",
    "        csv_content1 = open(input.m_csv).read()\n",
    "        /* more */\n",
    "\n",
    "```\n",
    "+ script:\n",
    ">\n",
    "> 放入要執行的python_code_file\n",
    ">\n",
    "\n",
    "```\n",
    "\n",
    "rule ebar_chart:\n",
    "    input:\n",
    "        \"resource/tp-economy.csv\"\n",
    "    output:\n",
    "        \"result/bar/e-bar.html\"\n",
    "    script:\n",
    "        \"scripts/make_bar_chart.py\"\n",
    "```\n",
    "\n",
    "+ log:\n",
    "> 輸出日誌文件的路徑\n",
    ">\n",
    "> 路徑名需加\" \"\n",
    ">\n",
    "> 在shell的指令候用 … 2> {log} 完成\n",
    ">\n",
    "\n",
    "+ params:\n",
    "> 傳遞額外的參數到 shell 指令或其他地方\n",
    ">\n",
    "> 以{param}使用\n",
    ">\n",
    "\n",
    "``` \n",
    "# toy example\n",
    "params:\n",
    "    extra_params=\"--option1 value1 --option2 value2\"\n",
    "shell:\n",
    "    \"process_data {input} {params.extra_params} > {output}\"\n",
    "```\n",
    "\n",
    "+ threads: 和resources: \n",
    ">  threads -> 規則所需的CPU數量\n",
    ">\n",
    "> resources -> 規則所需的如內存、磁碟空間...\n",
    ">\n",
    "\n",
    "### 配置文件 (config.yaml)\n",
    "\n",
    "```\n",
    "# config.yaml\n",
    "samples:\n",
    "    A: data/samples/A.fastq\n",
    "    B: data/samples/B.fastq\n",
    "    C: data/samples/C.fastq\n",
    "```\n",
    "\n",
    "```\n",
    "# Snakefile\n",
    "def get_bwa_map_input_fastqs(wildcards):\n",
    "    return config[\"samples\"][wildcards.sample]\n",
    "rule bwa_map:\n",
    "    input:\n",
    "        \"data/genome.fa\",\n",
    "        get_bwa_map_input_fastqs\n",
    "        # \"data/samples/{sample}.fastq\"\n",
    "    output:\n",
    "        \"mapped_reads/{sample}.bam\"\n",
    "    params:\n",
    "        rg=r\"@RG\\tID:{sample}\\tSM:{sample}\"\n",
    "    log:\n",
    "        \"logs/bwa_mem/{sample}.log\"\n",
    "    threads: 8\n",
    "    shell:\n",
    "        \"(bwa mem -R '{params.rg}' -t {threads} {input} | \"\n",
    "        \"samtools view -Sb - > {output}) 2> {log}\"\n",
    "\n",
    "\n",
    "configfile: \"config.yaml\" # config file 的路徑\n",
    "\t/..more../\n",
    "rule bcftools_call:\n",
    "    input:\n",
    "        fa=\"data/genome.fa\",\n",
    "        bam=expand(\"sorted_reads/{sample}.sorted.bam\", sample=config[\"samples\"]),\n",
    "        bai=expand(\"sorted_reads/{sample}.sorted.bam.bai\", sample=config[\"samples\"])\n",
    "/..more../\n",
    "```\n",
    "> + \"config[\"samples\"][wildcards.sample]\"取得 config 檔案裡sample下的路徑，在shell中使用{input}時會將三個路徑都當輸入跑一遍  \n",
    ">  \n",
    "> + {sample} 是通配符，會得到樣本名稱(**不是路徑**，所以```logs/bwa_mem/```裡面是A,B,C的log檔)\n",
    ">  \n",
    "> + ```bam=expand(\"sorted_reads/{sample}.sorted.bam\", sample=config[\"samples\"])```會讓 bam = sorted_reads/{A,B,C}.sorted.bam 三個檔案，在shell用 {input.bam} 會三個檔案都跑\n",
    ">  \n",
    "\n",
    "### 內建輸出\n",
    "+ 輸出工作流程圖(DAG有向無環圖)\n",
    "```snakemake –dag **/target/** | dot -Tsvg > **/file_output_name/**.svg```\n",
    "\n",
    "![dag](pictures/dag.png)\n",
    "\n",
    "+ 輸出report (內建workflow, statistics, about)\n",
    "```snakemake -s(op) **/target_snakefile/** –report **/report_name/** (.zip or .html …)``` \n",
    "\n",
    "![report-dag](pictures/report-dag.png)\n",
    "\n",
    "### Snakemake參數\n",
    "+ -n (--dry-run) : 模擬試跑workflow(for debug) ex: $ snakemake -n\n",
    "+ -p : 列出所有紀錄(for debug) \n",
    "ex: ```$ snakemake -p```\n",
    "+ -f (-F) : 強制重新跑完整workflow\n",
    "ex: ```$ snakemake -f```\n",
    "+ -n (-name): 指定名稱\n",
    "ex: ```$ conda env create -n example_env```\n",
    "+ –file: 指定安裝環境\n",
    "ex: ```$ conda env create -n example_env –file environment.yaml```\n",
    "+ -s : 指定跑的Snakefile \n",
    "ex: ```$ snakemake -s new_snakefile –report report.html```\n",
    "\n",
    "## III. 範例\n",
    "\n",
    "## 範例\t\n",
    "\n",
    "#### [EX1: Data analysis of saccharomyces cerevisiae ( Brewer's yeast)](https://pid.depositar.io/ark:37281/k56253r3p)\n",
    "\n",
    "(credit to [Catalog example](https://github.com/snakemake-workflows/rna-seq-star-deseq2))\n",
    "\n",
    "(1) 作者在depositar上資料集放入script,Snakefile,測資, workflow.zip(optional)\n",
    "\n",
    "![data](pictures/data.png)\n",
    "\n",
    "(2) 使用者在BinderHub中可將data.zip, sample, units更換為自己的資料，或是直接跑測試資料  \n",
    "![upload](pictures/upload.png)\n",
    "\n",
    "> 上傳資料  \n",
    "\n",
    "![rename_delete](pictures/rename_delete.png)\n",
    "\n",
    "> 重新命名或刪除    \n",
    "\n",
    "![terminal](pictures/terminal.png)  \n",
    "> 打開終端機  \n",
    "\n",
    "![bash](pictures/bash.png)  \n",
    "> 在終端機中輸入 $ bash script.sh  \n",
    "\n",
    "(3) Snakefile便會自動跑資料分析，並將結果從zip檔解壓縮成資料夾，使用者可在資料夾中的html檔中查看分析結果  \n",
    "\n",
    "![task](pictures/task_to_do.png)  \n",
    "> 秀出要執行的工作  \n",
    "\n",
    "![done](pictures/done.png)  \n",
    "> 100% done  \n",
    "\n",
    "![report](pictures/report.png)  \n",
    "> report的html檔  \n",
    "\n",
    "#### [EX2: 台北市金價與結婚人數分析](https://pid.depositar.io/ark:37281/k5b8f3n3j)(以前面為提及過的內容為主)\n",
    "\n",
    "> 利用而外的pyhton code得到金價與結婚對數的bar chart\n",
    "```\n",
    "# make_bar_chart.py\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "# Access the input and output files from Snakemake\n",
    "input_file = str(snakemake.input[0])\n",
    "output_file = str(snakemake.output[0])\n",
    "\n",
    "df = pd.read_csv(input_file)\n",
    "\n",
    "if 'marriage' in input_file:\n",
    "    col1 = '結婚對數/總計[對]'\n",
    "    col2 = '年別'\n",
    "    layout_title = '台北結婚對數變化圖'\n",
    "    ytitle = '結婚對數/總計[對]'\n",
    "    markcolor = 'blue'\n",
    "elif 'economy' in input_file:\n",
    "    col1 = '黃金[飾金]市價[元/臺錢]'\n",
    "    col2 = '年底別'\n",
    "    layout_title = '台北黃金市價變化圖'\n",
    "    ytitle = '黃金市價[元/臺錢]'\n",
    "    markcolor = 'gold'\n",
    "\n",
    "bar = go.Bar(\n",
    "    x=df[col2],\n",
    "    y=df[col1],\n",
    "    marker=dict(color=markcolor)\n",
    ")\n",
    "\n",
    "layout = go.Layout(\n",
    "    title=layout_title,\n",
    "    title_font_size = 30,\n",
    "    xaxis=dict(title=\"年分\"),\n",
    "    yaxis=dict(title=ytitle)\n",
    ")\n",
    "\n",
    "fig = go.Figure(data=bar, layout=layout)\n",
    "fig.write_html(output_file)\n",
    "```\n",
    "\n",
    "> 在snakefile裡產生html檔\n",
    "```\n",
    "# Snakefile\n",
    "\n",
    "\t/..more../\n",
    "rule mbar_chart:\n",
    "    input:\n",
    "        \"resource/tp-marriage.csv\"\n",
    "    output:\n",
    "        \"result/bar/m-bar.html\"\n",
    "    script:\n",
    "        \"scripts/make_bar_chart.py\"\n",
    "\n",
    "\t/..more../\n",
    "\n",
    "rule custom_report:\n",
    "    \n",
    "    /..more../\n",
    "\n",
    "    output:\n",
    "        \"new_report.html\",\n",
    "        \"dag_graph.svg\"\n",
    "    run:\n",
    "        # generate the DAG graph\n",
    "        shell(\"snakemake --dag | dot -Tsvg > {output[1]}\")\n",
    "\n",
    "        # read report content\n",
    "        with open(input.report) as f:\n",
    "            report_content = f.read()\n",
    "       \n",
    "        # read CSV file contents\n",
    "        with open(input.m_csv) as f1:\n",
    "            csv_content1 = f1.read()\n",
    "        with open(input.e_csv) as f2:\n",
    "            csv_content2 = f2.read()\n",
    "       \n",
    "        # generate the new HTML content\n",
    "        new_content = f\"\"\"\n",
    "        <!DOCTYPE html>\n",
    "        <html>\n",
    "        <head>\n",
    "            <title>Report of the Relation Between Economy and Marriage</title>\n",
    "        </head>\n",
    "        <body>\n",
    "            <h1>Snakemake Report</h1>\n",
    "\n",
    "           /..more../\n",
    "\n",
    "        </body>\n",
    "        </html>\n",
    "        \"\"\"\n",
    "       \n",
    "        # write the new report\n",
    "        with open(output[0], \"w\") as f:\n",
    "            f.write(new_content)\n",
    "\n",
    "```\n",
    "\n",
    "> output 出 三個檔案\n",
    "\n",
    "1. dag-graph  \n",
    "![dag-graph](pictures/dag-graph.png)\n",
    "\n",
    "2. new_report  \n",
    "![new_report](pictures/new_report.png)\n",
    "\n",
    "3. report  \n",
    "![old_report](pictures/default_report.png)\n",
    "\n",
    "github link: https://github.com/HutakiHare/2024AS-intern-snakemake/tree/778e4cb9a789f4eaef05ed3b4ebb866f7705828f/2024AS-intern-snakemake/%E7%AF%84%E4%BE%8B%E6%AA%94%E6%A1%88  \n",
    "\n",
    "depositar:  \n",
    "    + [snakemake paper report](https://pid.depositar.io/ark:37281/k5c4x9n7d)  \n",
    "    + [ex1:Data analysis of saccharomyces cerevisiae ( Brewer's yeast)](https://pid.depositar.io/ark:37281/k56253r3p)  \n",
    "    + [ex2: toy example-relation of gold price and marriage number](https://pid.depositar.io/ark:37281/k5b8f3n3j)  \n",
    "    \n",
    "webpage link: https://hutakihare.github.io/2027AS-intern-snakemake/"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}